packages/xl-ai/src/AIExtension.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  BlockNoteEditor,
+  BlockNoteExtension,
+  UnreachableCaseError,
+} from "@blocknote/core";
+import { LanguageModel } from "ai";
+import { createStore } from "zustand/vanilla";
+import { PromptOrMessages, llm } from "./api";
+import { LLMRequestOptions } from "./api/streamTool/callLLMWithStreamTools";
+
+type AIPluginState = {
+  aiMenuBlockID: string | undefined;
+  aiMenuResponseStatus: "initial" | "generating" | "error" | "done";
+};
+
+// parameters that are shared across all calls and can be configured on the context as "application wide" settings
+type GlobalLLMCallOptions = {
+  model: LanguageModel;
+  dataFormat?: "html" | "json" | "markdown";
+  stream?: boolean;
+};
+
+// parameters that are specific to each call
+type CallSpecificCallLLMOptions = Partial<GlobalLLMCallOptions> &
+  Omit<LLMRequestOptions, "messages" | "model"> &
+  PromptOrMessages & {
+    defaultStreamTools?: {
+      /** Enable the add tool (default: true) */
+      add?: boolean;
+      /** Enable the update tool (default: true) */
+      update?: boolean;
+      /** Enable the delete tool (default: true) */
+      delete?: boolean;
+    };
+  };
+
+// TO DISCUSS: extension vs plugin vs addon vs ..
+export class AIExtension extends BlockNoteExtension {
+  public static name(): string {
+    return "ai";
+  }
+
+  /**
+   * zustand design considerations
+   *
+   * - setters are not added in this store, but as class methods so we can have control over what to expose as API
+   * - con: if we'd expose the store externally, consumers could still call store.setState :/ (for this store it's not desired, for `options` below, it is)
+   *
+   */
+  public readonly store = createStore<AIPluginState>()((set) => ({
+    aiMenuBlockID: undefined,
+    aiMenuResponseStatus: "initial",
+  }));
+
+  /**
+   * zustand design considerations
+   *
+   * - I decided to create a separate store for the options as they feel quite different from "internal state"
+   *   above. They're not so much "state", but more "configuration". I still think it's useful to have in a Zustand
+   *   store (instead of a just a plain object), because consumers might want to display the settings in the UI.
+   *   Case can also be made to make this a plain and leave the responsibility of displaying settings to the consumer
+   *   (and not have a store for this at all). TO DISCUSS
+   */
+  public readonly options: ReturnType<
+    ReturnType<typeof createStore<Required<GlobalLLMCallOptions>>>
+  >;
```

Comment:
Yea, I think that configuration should just be a plain object, if the UI needs to be able to modify it, then the configur option should just be a default/initial value & the value is stored into the state store.

I'd imagine configuration like `codeBlock.supportedLanguages`, & I think it'd be complex to have that value be modifiable at runtime.

Reply:
I'd even go as far as to say that configuration should be `T extends Readonly<Record<string, any>>`

Reply:
not sure about this, let's discuss later (not zustand specific)